### PR TITLE
Read PackageVersion information for non-sdk projects

### LINKED ISSUE #9236: [Centrally managing NuGet packages and versions] Integrate PackageVersion information for non-sdk style projects in VS restore
For non sdk style scenarios the project system data comes from a different pipeline. Integrate PackageVersion information to support restore for projects that are opt-in CPVM.

Related with https://github.com/NuGet/Home/issues/8752

### PR DESCRIPTION
**Pull request by cristinamanum**
## Feature

Fixes: https://github.com/NuGet/Home/issues/9236 
Regression: No  
  

## Fix

Central package version management uses a new MsBuild item type ("PackageVersion"). Project System added new apis to support reading the PackageVersion items. 
This PR integrates the non-sdk PR projects with the new apis. The PackageVersion information is made available to the restore of VisualStudio projects that have the central package version management enabled. 

## Testing/Validation

Tests Added: Yes

Validation: Unit tests, functional tests and manual e2e.

### CODE REVIEW

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsManagedLanguagesProjectSystemServices.cs
Comment (zivkan) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsManagedLanguagesProjectSystemServices.cs]: It's unclear to me what the comment means. Is it a todo, that this should be modified to preserve current behaviour before being merged into the dev branch?  Or it's a todo that we might not fix until a future release? Or is it saying that the code works in this way in order to preserve current behaviour? (then what's the point of the url to the issue?)

If we keep the comment, please either re-open the linked issue, or change the URL to a open issue.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsManagedLanguagesProjectSystemServices.cs
Comment (zivkan) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsManagedLanguagesProjectSystemServices.cs]:
```suggestion
            if (isCpvmEnabled && string.IsNullOrEmpty(version))
            {
                // CPVM allows (expects) PackageReferences without versions
                return null;
            }

            return VersionRange.Parse(version);
```

File: src/NuGet.Clients/NuGet.VisualStudio.Common/IVsProjectAdapter.cs
Comment (zivkan) [inline src/NuGet.Clients/NuGet.VisualStudio.Common/IVsProjectAdapter.cs]: I think `IVsProjectAdapter` should be for methods needed to determine what the project type is (PC, legacy PR, CPS PR). So arguably the lock file methods shouldn't be in here, but at least that might be common across multiple project types. Adding these methods, which are specific to `LegacyProjectReferenceProject` does not feel correct here.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs
Comment (zivkan) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs]: `LegacyPackageReferenceProject` extends `BuildIntegratedNuGetProject` which itself extends `NuGetProject`, which has a `Metadata` property, so you can get the MSBuild property from an existing member, without adding new APIs to `IVsProjectAdapter`.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs
Comment (zivkan) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs]: Below you modified a call to `EnvDTEProjectUtility.GetProjectItemsAsync`. Can we use that, or something else that already exists in `LegacyPackageReferenceProject` instead of introducing a new API into `IVsProjectAdapter`?

File: src/NuGet.Clients/NuGet.VisualStudio.Common/VisualStudioBuildItemStorageCallback.cs
Comment (zivkan) [inline src/NuGet.Clients/NuGet.VisualStudio.Common/VisualStudioBuildItemStorageCallback.cs]: Static instances only work for stateless or immutable objects. This object's `Items` property gets modified, which means if you have multiple legacy PR projects, or you unload and reload a single legacy PR project, bad things will happen.

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/ProjectSystems/LegacyPackageReferenceProjectTests.cs
Comment (zivkan) [inline test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/ProjectSystems/LegacyPackageReferenceProjectTests.cs]: I think `packageAv1` and `packageAv5` would be easier to understand. `packageAA` makes it sound like it's a different package to `packageA`.

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/ProjectSystems/LegacyPackageReferenceProjectTests.cs
Comment (zivkan) [inline test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/ProjectSystems/LegacyPackageReferenceProjectTests.cs]:
```suggestion
        public async Task GetPackageSpecAsync_CentralPackageVersionsRemovedDuplicates()
```

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/ProjectSystems/LegacyPackageReferenceRestoreUtilityTests.cs
Comment (zivkan) [inline test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/ProjectSystems/LegacyPackageReferenceRestoreUtilityTests.cs]: Are the targets and props files relevant to the test? I think it could be simplified a little if removed. same for the other package below.

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/ProjectSystems/LegacyPackageReferenceRestoreUtilityTests.cs
Comment (zivkan) [inline test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/ProjectSystems/LegacyPackageReferenceRestoreUtilityTests.cs]: Can you name this argument, please. Without looking at the definition of `RestoreAsync`, I have no idea why a empty anonymous function is being passed in.  Also consider using a discard `_` instead of an unused named variable `c`.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsManagedLanguagesProjectSystemServices.cs
Comment (cristinamanum) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsManagedLanguagesProjectSystemServices.cs]: This line will throw if the version is empty but this issue is already existent and the current change will not attempt to fix it. The linked issue was closed as a duplicate. Removed the reference to the bug as it seems it is part of a larger problem.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsManagedLanguagesProjectSystemServices.cs
Comment (cristinamanum) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsManagedLanguagesProjectSystemServices.cs]: Updated

File: src/NuGet.Clients/NuGet.VisualStudio.Common/IVsProjectAdapter.cs
Comment (cristinamanum) [reply src/NuGet.Clients/NuGet.VisualStudio.Common/IVsProjectAdapter.cs]: Following the description of the `IVsProjectAdapter` [here](https://github.com/NuGet/NuGet.Client/blob/818ca73c9d788fa7afe05567d16b0e84dd88edca/src/NuGet.Clients/NuGet.VisualStudio.Common/IVsProjectAdapter.cs#L15) the IVsProjectAdapter will access a project properties. Started an offline discussion to agree on the right location of these apis.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs
Comment (cristinamanum) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs]: One comment above.

File: src/NuGet.Clients/NuGet.VisualStudio.Common/VisualStudioBuildItemStorageCallback.cs
Comment (cristinamanum) [reply src/NuGet.Clients/NuGet.VisualStudio.Common/VisualStudioBuildItemStorageCallback.cs]: Discussed offline. There is not an issue with the functionality. However it is hard to read, so I refactored the code.

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/ProjectSystems/LegacyPackageReferenceProjectTests.cs
Comment (cristinamanum) [reply test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/ProjectSystems/LegacyPackageReferenceProjectTests.cs]: Changed the names.

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/ProjectSystems/LegacyPackageReferenceProjectTests.cs
Comment (cristinamanum) [reply test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/ProjectSystems/LegacyPackageReferenceProjectTests.cs]: Changed the name.

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/ProjectSystems/LegacyPackageReferenceRestoreUtilityTests.cs
Comment (cristinamanum) [reply test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/ProjectSystems/LegacyPackageReferenceRestoreUtilityTests.cs]: Updated.

File: src/NuGet.Clients/NuGet.VisualStudio.Common/IVsProjectAdapter.cs
Comment (cristinamanum) [reply src/NuGet.Clients/NuGet.VisualStudio.Common/IVsProjectAdapter.cs]: Updated the  `IVsProjectAdapter` to be agnostic to specific items/property to read.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs
Comment (cristinamanum) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs]: Please see the above comment.

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/ProjectSystems/LegacyPackageReferenceRestoreUtilityTests.cs
Comment (cristinamanum) [reply test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/ProjectSystems/LegacyPackageReferenceRestoreUtilityTests.cs]: No - removed.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsManagedLanguagesProjectSystemServices.cs
Comment (zivkan) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsManagedLanguagesProjectSystemServices.cs]: The value of the MSBuild item is not expected to change, even if someone refactors code and changes the name of `ProjectBuildProperties.ManagePackageVersionsCentrally`.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs
Comment (zivkan) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs]: The value of the MSBuild item is not expected to change, even if someone refactors code and changes the name of `ProjectBuildProperties.ManagePackageVersionsCentrally`.

File: src/NuGet.Clients/NuGet.VisualStudio.Common/IVsProjectAdapter.cs
Comment (zivkan) [reply src/NuGet.Clients/NuGet.VisualStudio.Common/IVsProjectAdapter.cs]: yes, it appears I misunderstood the purpose of `IVsProjectAdapter`. The first time I saw it was when I was trying to figure out how NuGet determines which project type (PC vs legacy PR vs CPS PR) a project uses, so in my head, I associated this interface with only needing to figure out how to get a more specific project adapter. Appears I was wrong 😢

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsManagedLanguagesProjectSystemServices.cs
Comment (cristinamanum) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsManagedLanguagesProjectSystemServices.cs]: I followed the pattern used for other properties. Is the suggestion to make it constant?

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs
Comment (cristinamanum) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs]: I followed the pattern used for other properties. Is the suggestion to make it constant?

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsManagedLanguagesProjectSystemServices.cs
Comment (zivkan) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsManagedLanguagesProjectSystemServices.cs]: When I search for other usages of `ProjectBuildProperties.`, I don't see any of them using `nameof`. I dislike that the fields defined in that class use the `field = nameof(field)`, but that's different to using `ProjectBuildProperties.ManagePackageVersionsCentrally` here (`Method(ProjectBuildProperties.ManagePackageVersionsCentrally)` vs (`Method(nameof(ProjectBuildProperties.ManagePackageVersionsCentrally))`).

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs]: The pattern that we are using is the following: 

```cs
public static class ProjectBuildProperties {
        public const string ManagePackageVersionsCentrally= nameof(ManagePackageVersionsCentrally);
}
```

Then later you just use: 

```cs
return MSBuildStringUtility.IsTrue(await _vsProjectAdapter.GetPropertyValueAsync(ProjectBuildProperties.ManagePackageVersionsCentrally));

```

Comment (nkolev92) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs]: I disagree about the refactoring concern. 
I think it's not a concern. The idea with nameof is to avoid typos everywhere and keep things consistent. 

An engineer should be aware that changing a property can technically break the scenario.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsManagedLanguagesProjectSystemServices.cs
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsManagedLanguagesProjectSystemServices.cs]: Refer to: https://github.com/NuGet/NuGet.Client/pull/3344/files#r412549827

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs]: style: too many empty lines here :)

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/VsProjectAdapter.cs
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/VsProjectAdapter.cs]:
```suggestion
        public Task<string> GetNuGetLockFilePathAsync()
        {
            return GetPropertyValueAsync(ProjectBuildProperties.NuGetLockFilePath);
        }
```

Saves some allocations.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/VsProjectAdapter.cs
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/VsProjectAdapter.cs]: Use param, you don't need the list.

File: src/NuGet.Clients/NuGet.VisualStudio.Common/IVsProjectAdapter.cs
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.VisualStudio.Common/IVsProjectAdapter.cs]: Use params.

File: src/NuGet.Clients/NuGet.VisualStudio.Common/NuGet.VisualStudio.Common.csproj
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.VisualStudio.Common/NuGet.VisualStudio.Common.csproj]: You are missing a space at the end of the xml line.

Comment (nkolev92) [inline src/NuGet.Clients/NuGet.VisualStudio.Common/NuGet.VisualStudio.Common.csproj]:
```suggestion
    <PackageReference Include="Microsoft.VisualStudio.Shell.Interop.16.0.DesignTime" />
    <PackageReference Include="Microsoft.VisualStudio.Shell.Interop.16.6.DesignTime" />
```

**Review by nkolev92 — COMMENTED** [review]
Looks great at this point. 
Mostly minor things.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs
Comment (zivkan) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs]: > The idea with nameof is to avoid typos everywhere and keep things consistent.

Avoiding typos and keeping things consistent is achieved by having the single property that all code re-uses, rather than defining their own local value. `nameof` doesn't solve that.

> An engineer should be aware that changing a property can technically break the scenario.

I find that really scary. Renaming a variable should never break code.

Comment (nkolev92) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs]: > I find that really scary. Renaming a variable should never break code.

I have never seen issues caused because of that. I think it makes an assumption that people don't pay any attention to what's being committed :) 

It's common in many repos: https://github.com/dotnet/project-system/blob/master/src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/ProjectCapability.cs
https://github.com/dotnet/sdk/blob/master/src/Cli/dotnet/UILanguageOverride.cs

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs
Comment (cristinamanum) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs]: Removed empty line.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/VsProjectAdapter.cs
Comment (cristinamanum) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/VsProjectAdapter.cs]: Updated

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/VsProjectAdapter.cs
Comment (cristinamanum) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/VsProjectAdapter.cs]: Updated.

File: src/NuGet.Clients/NuGet.VisualStudio.Common/IVsProjectAdapter.cs
Comment (cristinamanum) [reply src/NuGet.Clients/NuGet.VisualStudio.Common/IVsProjectAdapter.cs]: Updated.

File: src/NuGet.Clients/NuGet.VisualStudio.Common/NuGet.VisualStudio.Common.csproj
Comment (cristinamanum) [reply src/NuGet.Clients/NuGet.VisualStudio.Common/NuGet.VisualStudio.Common.csproj]: Updated.

Comment (cristinamanum) [reply src/NuGet.Clients/NuGet.VisualStudio.Common/NuGet.VisualStudio.Common.csproj]: Updated

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs
Comment (cristinamanum) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs]: Updated with _vsProjectAdapter.GetPropertyValueAsync(ProjectBuildProperties.ManagePackageVersionsCentrally)

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsManagedLanguagesProjectSystemServices.cs
Comment (cristinamanum) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsManagedLanguagesProjectSystemServices.cs]: Updated with _vsProjectAdapter.GetPropertyValueAsync(ProjectBuildProperties.ManagePackageVersionsCentrally)

**Review by nkolev92 — APPROVED** [review]
LGTM. 

I'd wait on @zivkan's review before merging.

@cristinamanum Consider adding a manual test for the test team given that we're unable to test this E2E because our machines are not all up to date.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs
Comment (zivkan) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs]: If you don't need to make any other changes & run a new build, don't bother changing this, but please try to be careful to avoid these types of unnecessary whitespace changes.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs
Comment (zivkan) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs]: Both `PackageVersion` and `Version` are defined in `PackageBuildProperties`.

**Review by zivkan — APPROVED** [review]

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs
Comment (cristinamanum) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs]: Updated

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs
Comment (cristinamanum) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs]: Updated

**Pull request merged**
